### ISSUE TITLE
Refer a Keras layer in a layer.

### ISSUE THREAD

**Issue by duguyue100** [fchollet/keras#5839]
This is rather a general question. Just for the sake of simplicity and not re-inventing the wheel, is there a good way of refer another layer in a layer?

Say I have more than one tensor to compute in call function, and I want to deploy BatchNormalization to both of them, this means I need a in-built batch normalization in there.

I sort of come out one ugly solution where I simply replicated the entire BatchNormalization layer and made some tricks so it appears to be working.

**Comment by PavlosMelissinos** [comment]
I'm not confident I understand exactly what your goal is but I think you'd be interested in [shared layers](https://keras.io/getting-started/functional-api-guide/#shared-layers). The most straightforward way - that I can think of - to reuse weights within complex layers would probably be to use your shared layer (BatchNormalization) and your normal input as inputs to a merge layer.

Otherwise, since this looks like an implementation question, have you tried posting your question on [stack overflow](http://stackoverflow.com/questions/tagged/keras) or [the slack channel](https://keras-slack-autojoin.herokuapp.com/)? **The issue tracker is supposed to be about bug reports and feature requests, it's not meant to be a q&a forum.**

From the "New issue" form:

> If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

From contribute.md

> Your code doesn't work, and you have determined that the issue lies with Keras? Follow these steps to report a bug.

> You can also use Github issues to request features you would like to see in Keras, or changes in the Keras API.

We should all try to keep it as clean as possible and help the developers.

Cheers.

**Comment by duguyue100** [comment]
Hi, @PavlosMelissinos, thanks for the answer. I don't think it's quite fit my needs. FYI, I tried with StackOverflow, but no feedback yet. And thanks for tolerating me here.

What I really want is like in this form, say I'm defining a new recurrent layer.

```python
class FooLayer(Recurrent):
    def __init__(self, **kwargs):
        # some line of codes...
        self.b1 = BatchNormalization(axis)
        self.b2 = BatchNormalization(axis)
    def call(x, htm1):
        h = self.b1(x)+self.b2(htm1)+self.bias
        h = self.activation(h)
        return h, [h]
```

I hope I can embed other layers in my custom layer. And as for recurrent network, I need to report hidden states also, I'm not quite sure shared layer could do the job.

Or are you saying I can initialise my batch normalisation layers out there and pass as an argument?

**Comment by bstriner** [comment]
In general, you can use layers within other layers. In your specific example of using BN in a recurrent layer, you really shouldn't. Parameters are handled differently within recurrent calls than within normal layers. 

Code like what you have should still work but won't optimize correctly because `scan` (what does recurrence) doesn't know about the batchnorm parameters. Alternatively, use a BN mode that doesn't have parameters or copy the BN.call code into your step function.

If you implement a custom layer using code from BN, you can make sure the mean and sigma are passed to `scan` and the whole thing will optimize correctly.

If all you want to do is have BN in your own layer, try `keras.backend.batch_normalization`. You don't really need the BN layer to just call the BN function and it will optimize the best in an RNN.

https://keras.io/backend/

**Comment by duguyue100** [comment]
Hi, @bstriner, do you mean I can model BN function like LSTM's memory cell, a customised function and simply pass the states in return? I think this is indeed the path I'm following.

There is one more question though, in BN code, there is one update function:

```python
        if training in {0, False}:
            return normed
        else:
            self.add_update([K.moving_average_update(self.moving_mean,
                                                     mean,
                                                     self.momentum),
                             K.moving_average_update(self.moving_variance,
                                                     variance,
                                                     self.momentum)],
                            inputs)
```

Is this line simply saying the states of moving mean and variance depends on inputs in gradient updates?
Apparently when I run code with this part it doesn't work, this should be the state I return in the step function right?

**Comment by bstriner** [comment]
One mode is to BN the test data using the test data. Another mode is to BN the test data using parameters learned during training. 
* If you do the first mode, you can use the BN layer directly or just call K.batch_normalization. 
* If you want to do the second mode within an RNN, you need to add those parameters to your layer and to self.constants then get those parameters from the arguments to your step function and then use them correctly.
You can probably get away with just using BN in the first mode and if you run into inconsistencies you can implement the moving averages.
